src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs

```diff
@@ -96,35 +97,37 @@ protected LineMappingEntry FindEntry(int lineNumber)
         // Find the index of the line mapped entry with the largest unmapped line number <= lineNumber.
         protected int FindEntryIndex(int lineNumber)
         {
-            int r = Array.BinarySearch(this.Entries, new LineMappingEntry(lineNumber));
+            int r = ImmutableArray.BinarySearch(this.Entries, new LineMappingEntry(lineNumber));
             return r >= 0 ? r : ((~r) - 1);
         }
 
         // Given the ordered list of all directives in the file, return the ordered line mapping
         // entry for the file. This always starts with the null mapped that maps line 0 to line 0.
-        private LineMappingEntry[] CreateEntryMap(SourceText sourceText, IEnumerable<TDirective> directives)
+        private ImmutableArray<LineMappingEntry> CreateEntryMap(SyntaxTree tree, IList<TDirective> directives)
         {
-            var entries = new LineMappingEntry[directives.Count() + 1];
+            var entries = ArrayBuilder<LineMappingEntry>.GetInstance(directives.Count + 1);
             var current = InitializeFirstEntry();
-            var index = 0;
-            entries[index] = current;
+            entries.Add(current);
 
-            foreach (var directive in directives)
+            if (directives.Count > 0)
             {
-                current = GetEntry(directive, sourceText, current);
-                ++index;
-                entries[index] = current;
+                var sourceText = tree.GetText();
```

Comment:
HasHiddenRegions is a public API, so I feel we should keep the current approach and add a public API with cancellation token if we see performance delays.